File: src/main/java/com/jore/epoc/bo/orders/BuyRawMaterialOrder.java
package com.jore.epoc.bo.orders;

import org.hibernate.annotations.CompositeType;

import com.jore.Assert;
import com.jore.datatypes.money.Money;
import com.jore.epoc.bo.Storage;
import com.jore.epoc.bo.accounting.FinancialAccounting;
import com.jore.epoc.bo.message.MessageLevel;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;

@Entity
public class BuyRawMaterialOrder extends AbstractSimulationOrder {
    private Integer amount;
    @AttributeOverride(name = "amount", column = @Column(name = "unit_price_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "unit_price_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money unitPrice;

    @Override
    public void execute() {
        Assert.notNull("Amount must not be null", amount);
        Assert.notNull("Unit price must not be null", unitPrice);
        int storageCapacity = getCompany().getStorages().stream().mapToInt(storage -> storage.getAvailableCapacity(getExecutionMonth())).sum();
        if (storageCapacity >= amount && getCompany().getAccounting().checkFunds(getAmount(), getExecutionMonth().atEndOfMonth())) {
            Storage.distributeRawMaterialAccrossStorages(getCompany().getStorages(), amount, getExecutionMonth(), unitPrice);
            book(getExecutionMonth().atDay(FIRST_OF_MONTH), "Buy of raw material", FinancialAccounting.MATERIALAUFWAND, FinancialAccounting.BANK, getAmount(), FinancialAccounting.RAW_MATERIALS, FinancialAccounting.BESTANDESAENDERUNGEN_ROHWAREN, getAmount());
            addMessage(MessageLevel.INFORMATION, "RawMaterialBought", amount, getExecutionMonth());
            setExecuted(true);
        } else {
            if (!getCompany().getAccounting().checkFunds(getAmount(), getExecutionMonth().atEndOfMonth())) {
                addMessage(MessageLevel.WARNING, "NoRawMaterialFunds", getExecutionMonth(), getAmount(), getCompany().getAccounting().getBankBalance(getExecutionMonth().atEndOfMonth()));
            } else {
                addMessage(MessageLevel.WARNING, "NoRawMaterialCapacity", getExecutionMonth(), amount, storageCapacity);
            }
        }
    }

    @Override
    public Money getAmount() {
        return unitPrice.multiply(amount);
    }

    @Override
    public int getSortOrder() {
        return 3;
    }

    @Override
    public String getType() {
        return "Buy raw material";
    }

    public void setAmount(Integer amount) {
        this.amount = amount;
    }

    public void setUnitPrice(Money unitPrice) {
        this.unitPrice = unitPrice;
    }
}

File: EpocServer/Sources/App/Models/Orders/BuyRawMaterialOrder.swift
import Foundation

public final class BuyRawMaterialOrder: SimulationOrder {
    public var id: Int64?
    public var executionMonth: YearMonth = .of(2020, 1)
    public var isExecuted: Bool = false
    public weak var company: Company!
    public var amount: Int = 0
    public var unitPrice: Money = Money.of("CHF", 300)

    public var sortOrder: Int { 3 }
    public var type: String { "Buy raw material" }

    public func getAmount() -> Money {
        unitPrice.multiply(amount)
    }

    public func execute() {
        let storageCapacity = company.storages.reduce(0) { $0 + $1.getAvailableCapacity(executionMonth) }
        if storageCapacity >= amount && company.accounting.checkFunds(getAmount(), valueDate: executionMonth.atEndOfMonth()) {
            Storage.distributeRawMaterialAcrossStorages(company.storages, amount: amount, month: executionMonth, unitPrice: unitPrice)
            book(bookingDate: executionMonth.atDay(1), bookingText: "Buy of raw material",
                 debitAccount1: FinancialAccounting.MATERIALAUFWAND, creditAccount1: FinancialAccounting.BANK, amount1: getAmount(),
                 debitAccount2: FinancialAccounting.RAW_MATERIALS, creditAccount2: FinancialAccounting.BESTANDESAENDERUNGEN_ROHWAREN, amount2: getAmount())
            addMessage(.information, "RawMaterialBought", amount, executionMonth)
            isExecuted = true
        } else {
            if !company.accounting.checkFunds(getAmount(), valueDate: executionMonth.atEndOfMonth()) {
                addMessage(.warning, "NoRawMaterialFunds", executionMonth, getAmount(), company.accounting.getBankBalance(executionMonth.atEndOfMonth()))
            } else {
                addMessage(.warning, "NoRawMaterialCapacity", executionMonth, amount, storageCapacity)
            }
        }
    }
}
